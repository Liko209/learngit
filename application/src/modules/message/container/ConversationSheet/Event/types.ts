/*
 * @Author: Nello Huang (nello.huang@ringcentral.com)
 * @Date: 2018-11-07 15:47:16
 * Copyright Â© RingCentral. All rights reserved.
 */
import { PromisedComputedValue } from 'computed-async-mobx';
import EventItemModel from '@/store/models/EventItem';
import { Palette } from 'jui/foundation/theme/theme';

type ActivityData = {
  [index: string]: any;
};

type EventProps = {
  ids: number[];
};

type EventUpdateProps = {
  postId: number;
} & EventProps;

type EventViewProps = {
  event: EventItemModel;
  color?: [keyof Palette, string];
};

type EventUpdateViewProps = {
  activityData: ActivityData;
  event: EventItemModel;
  oldTimeText: PromisedComputedValue<string>;
  newTimeText: PromisedComputedValue<string>;
  color?: [keyof Palette, string];
};

export { EventProps, EventViewProps, EventUpdateViewProps, EventUpdateProps };
