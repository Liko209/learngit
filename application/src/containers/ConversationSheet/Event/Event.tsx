/*
 * @Author: Nello Huang (nello.huang@ringcentral.com)
 * @Date: 2018-11-07 14:52:23
 * Copyright Â© RingCentral. All rights reserved.
 */
import { buildContainer } from '@/base';
import { EventView } from './Event.View';
import { EventViewModel } from './Event.ViewModel';
import { EventProps } from './types';

const Event = buildContainer<EventProps>({
  ViewModel: EventViewModel,
  View: EventView,
});

export { Event };
