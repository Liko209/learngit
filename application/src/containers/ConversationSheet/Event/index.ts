/*
 * @Author: Nello Huang (nello.huang@ringcentral.com)
 * @Date: 2018-11-07 14:45:37
 * Copyright Â© RingCentral. All rights reserved.
 */
export * from './Event';
