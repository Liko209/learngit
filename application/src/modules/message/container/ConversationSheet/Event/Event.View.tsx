/*
 * @Author: Nello Huang (nello.huang@ringcentral.com)
 * @Date: 2018-11-07 15:37:47
 * Copyright Â© RingCentral. All rights reserved.
 */
import React from 'react';
import { observer } from 'mobx-react';
import { withTranslation, WithTranslation } from 'react-i18next'; // use external instead of injected due to incompatible with SortableElement
import { JuiConversationItemCard } from 'jui/pattern/ConversationItemCard';
import {
  JuiLabelWithContent,
  JuiEventDescription,
  JuiEventLocation,
  JuiTimeMessage,
  JuiSectionDivider,
} from 'jui/pattern/ConversationItemCard/ConversationItemCardBody';
import { EventViewProps } from './types';
import {
  postParser,
  HighlightContextInfo,
  SearchHighlightContext,
} from '@/common/postParser';
type Props = WithTranslation & EventViewProps;

@observer
class Event extends React.Component<Props, {}> {
  static contextType = SearchHighlightContext;
  context: HighlightContextInfo;
  render() {
    const { event, t, color, timeContent } = this.props;
    const { text, description, location } = event;

    return (
      <JuiConversationItemCard
        title={postParser(text, { keyword: this.context.keyword })}
        iconColor={color}
        titleColor={color}
        Icon="event"
      >
        <JuiSectionDivider gap={2}>
          <JuiLabelWithContent label={t('item.due')}>
            <JuiTimeMessage time={`${timeContent.get()}`} />
          </JuiLabelWithContent>
          {location && (
            <JuiLabelWithContent label={t('item.locationTitle')}>
              <JuiEventLocation>
                {postParser(location, {
                  keyword: this.context.keyword,
                  url: true,
                })}
              </JuiEventLocation>
            </JuiLabelWithContent>
          )}
          {description && (
            <JuiEventDescription>
              {postParser(description, {
                keyword: this.context.keyword,
                phoneNumber: true,
                url: true,
              })}
            </JuiEventDescription>
          )}
        </JuiSectionDivider>
      </JuiConversationItemCard>
    );
  }
}

const EventView = withTranslation('translations')(Event);

export { EventView };
