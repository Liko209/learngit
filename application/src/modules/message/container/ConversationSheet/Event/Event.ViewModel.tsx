/*
 * @Author: Nello Huang (nello.huang@ringcentral.com)
 * @Date: 2018-11-07 15:45:04
 * Copyright Â© RingCentral. All rights reserved.
 */
import { computed } from 'mobx';
import { StoreViewModel } from '@/store/ViewModel';
import { Item } from 'sdk/module/item/entity';
import { getEntity } from '@/store/utils';
import { ENTITY_NAME } from '@/store';
import { EventViewProps, EventProps } from './types';
import EventItemModel from '@/store/models/EventItem';
import { accentColor } from '@/common/AccentColor';

class EventViewModel extends StoreViewModel<EventProps>
  implements EventViewProps {
  @computed
  get _id() {
    return this.props.ids[0];
  }

  @computed
  get event() {
    return getEntity<Item, EventItemModel>(ENTITY_NAME.ITEM, this._id);
  }

  @computed
  get color() {
    return accentColor[this.event.color];
  }
}

export { EventViewModel };
