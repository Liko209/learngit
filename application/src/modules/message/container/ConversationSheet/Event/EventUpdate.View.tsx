/*
 * @Author: Nello Huang (nello.huang@ringcentral.com)
 * @Date: 2018-11-07 15:37:47
 * Copyright Â© RingCentral. All rights reserved.
 */
import React from 'react';
import { observer } from 'mobx-react';
import { withTranslation, WithTranslation } from 'react-i18next'; // use external instead of injected due to incompatible with SortableElement
import { JuiConversationItemCard as EventUpdateViewCard } from 'jui/pattern/ConversationItemCard';
import {
  JuiLabelWithContent,
  JuiEventLocation,
  JuiTimeMessage,
} from 'jui/pattern/ConversationItemCard/ConversationItemCardBody';
import {
  JuiEventCollapse,
  JuiEventCollapseContent,
} from 'jui/pattern/ConversationItemCard/ConversationItemCardFooter';
import { EventUpdateViewProps } from './types';

type Props = WithTranslation & EventUpdateViewProps;
@observer
class EventUpdate extends React.Component<Props> {
  private _isShowTime = (value: any) => {
    return (
      value.start ||
      value.end ||
      value.repeat ||
      value.repeat_ending_after ||
      value.repeat_ending_on ||
      value.repeat_ending
    );
  }

  private _getLocation = (value: any) => value.location;

  render() {
    const {
      oldTime,
      newTime,
      event,
      activityData,
      oldTimeText,
      newTimeText,
      color,
      t,
    } = this.props;
    const { text } = event;
    const { old_values, new_values } = activityData;

    const oldLocation = this._getLocation(old_values);
    const hasOldTime = this._isShowTime(old_values);

    const newLocation = this._getLocation(new_values);
    const hasNewTime = this._isShowTime(new_values);

    return (
      <EventUpdateViewCard
        title={text}
        titleColor={color}
        iconColor={color}
        Icon="event"
        Footer={
          (hasOldTime || oldLocation) && (
            <JuiEventCollapse
              showText={t('item.showEventHistory')}
              hideText={t('item.hideEventHistory')}
            >
              {hasOldTime && (
                <JuiEventCollapseContent noHighlight={true}>
                  {`${oldTime.get()} ${oldTimeText.get()}`}
                </JuiEventCollapseContent>
              )}
              {oldLocation && (
                <JuiEventCollapseContent>{oldLocation}</JuiEventCollapseContent>
              )}
            </JuiEventCollapse>
          )
        }
      >
        {hasNewTime && (
          <JuiLabelWithContent label={t('item.due')}>
            <JuiTimeMessage time={`${newTime.get()} ${newTimeText.get()}`} />
          </JuiLabelWithContent>
        )}
        {newLocation && (
          <JuiLabelWithContent label={t('item.locationTitle')}>
            <JuiEventLocation location={newLocation} />
          </JuiLabelWithContent>
        )}
      </EventUpdateViewCard>
    );
  }
}

const EventUpdateView = withTranslation('translations')(EventUpdate);

export { EventUpdateView };
