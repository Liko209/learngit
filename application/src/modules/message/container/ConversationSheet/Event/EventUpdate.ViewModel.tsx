/*
 * @Author: Nello Huang (nello.huang@ringcentral.com)
 * @Date: 2018-11-07 15:45:04
 * Copyright Â© RingCentral. All rights reserved.
 */
import { promisedComputed } from 'computed-async-mobx';
import { computed } from 'mobx';
import { StoreViewModel } from '@/store/ViewModel';
import { Item } from 'sdk/module/item/entity';
import { Post } from 'sdk/module/post/entity';
import PostModel from '@/store/models/Post';
import { getEntity } from '@/store/utils';
import { ENTITY_NAME } from '@/store';
import { EventUpdateViewProps, EventUpdateProps } from './types';
import EventItemModel from '@/store/models/EventItem';
import { getDurationTimeText } from '../helper';
import { accentColor } from '@/common/AccentColor';

class EventUpdateViewModel extends StoreViewModel<EventUpdateProps>
  implements EventUpdateViewProps {
  @computed
  get _id() {
    return this.props.ids[0];
  }

  @computed
  get _postId() {
    return this.props.postId;
  }

  @computed
  get post() {
    return getEntity<Post, PostModel>(ENTITY_NAME.POST, this._postId);
  }

  @computed
  get event() {
    return getEntity<Item, EventItemModel>(ENTITY_NAME.ITEM, this._id);
  }

  @computed
  get color() {
    return accentColor[this.event.color];
  }

  @computed
  get activityData() {
    return this.post.activityData || {};
  }

  getTimeText(value: any, event: EventItemModel) {
    const {
      repeat,
      repeatEndingAfter,
      repeatEnding,
      repeatEndingOn,
    } = this.event;
    return (
      (repeatEndingAfter &&
        getDurationTimeText(
          value.repeat || repeat,
          value.repeat_ending_after || repeatEndingAfter,
          value.repeat_ending_on || repeatEndingOn,
          value.repeat_ending || repeatEnding,
        )) ||
      ''
    );
  }

  oldTimeText = promisedComputed('', async () => {
    const { old_values } = this.activityData;
    return await this.getTimeText(old_values, this.event);
  });

  newTimeText = promisedComputed('', async () => {
    const { new_values } = this.activityData;
    return await this.getTimeText(new_values, this.event);
  });
}

export { EventUpdateViewModel };
