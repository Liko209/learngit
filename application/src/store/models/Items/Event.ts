/*
 * @Author: Nello Huang (nello.huang@ringcentral.com)
 * @Date: 2018-11-08 11:28:04
 * Copyright Â© RingCentral. All rights reserved.
 */
import { EventItem } from './';

function setEventData(this: EventItem, data: any) {
  const { color, description, start, end, location, repeat, repeat_ending, repeat_ending_after, repeat_ending_on, text } = data;

  this.color = color;
  this.description = description;
  this.start = start;
  this.end = end;
  this.location = location;
  this.repeat = repeat;
  this.repeat_ending = repeat_ending;
  this.repeat_ending_after = repeat_ending_after;
  this.repeat_ending_on = repeat_ending_on;
  this.text = text;
}

export { setEventData };
