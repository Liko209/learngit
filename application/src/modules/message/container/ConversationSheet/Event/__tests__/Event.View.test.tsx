import React from 'react';
import { shallow } from 'enzyme';
import { EventView } from '../Event.View';

describe('render()', () => {
  const props = {
    event: {
      location: 'location',
      text: 'text',
      description: 'description',
    },
    t: (e: any) => e,
    color: 'color',
    timeContent: {
      get: () => '1997/10/17',
    },
  } as any;
  it('render correctly', () => {
    const Wrapper = shallow(<EventView {...props} />);
    expect(Wrapper).toMatchSnapshot();
  });
});
