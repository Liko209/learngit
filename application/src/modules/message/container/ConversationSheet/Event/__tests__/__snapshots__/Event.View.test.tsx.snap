// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() render correctly 1`] = `
<ForwardRef
  Icon="event"
  iconColor="color"
  title="text"
  titleColor="color"
>
  <JuiSectionDivider
    gap={2}
  >
    <JuiLabelWithContent
      label="item.due"
    >
      <JuiTimeMessage
        time="1997/10/17"
      />
    </JuiLabelWithContent>
    <JuiLabelWithContent
      label="item.locationTitle"
    >
      <ForwardRef
        location="location"
      />
    </JuiLabelWithContent>
    <ComponentWithLink
      description="description"
    />
  </JuiSectionDivider>
</ForwardRef>
`;
